src/core/mac/mac.cpp

```diff
@@ -250,8 +232,54 @@ otError Mac::ConvertBeaconToActiveScanResult(Frame *aBeaconFrame, otActiveScanRe
     return error;
 }
 
-void Mac::StartEnergyScan(void)
+otError Mac::UpdateScanChannel(void)
+{
+    otError error = OT_ERROR_NONE;
+
+    do
+    {
+        mScanChannel++;
+        VerifyOrExit(mScanChannel <= OT_RADIO_CHANNEL_MAX, error = OT_ERROR_NOT_FOUND);
+
+    }
+    while ((mScanChannels & (1U << mScanChannels)) == 0);
+
+exit:
+    return error;
+}
+
+void Mac::PerformActiveScan(void)
+{
+    if (UpdateScanChannel() == OT_ERROR_NONE)
+    {
+        // If there are more channels to scan, start CSMA backoff to send the beacon request.
+        StartCsmaBackoff();
+    }
+    else
+    {
+        otPlatRadioSetPanId(&GetInstance(), mPanId);
+        mActiveScanHandler(mScanContext, NULL);
+        FinishOperation();
+    }
+}
+
+void Mac::PerformEnergyScan(void)
 {
+    otError error = OT_ERROR_NONE;
+
+    if (mScanDuration == 0)
+    {
+        while ((error = UpdateScanChannel()) == OT_ERROR_NONE)
+        {
+            RadioReceive(mScanChannel);
+            ReportEnergyScanResult(otPlatRadioGetRssi(&GetInstance()));
```

Comment:
These callbacks may occur before `Mac::EnergyScan()` returns.  Do we want to delay them to a following task?

Reply:
Good point. Will change this to be done from the task instead.